pkg/controllers/subnetbinding/subnetbinding_controller.go

```diff
@@ -203,52 +202,127 @@ func (r *Reconciler) listBindingMapIDsFromCRs(ctx context.Context) (sets.Set[str
+				message: fmt.Sprintf("Target Subnet %s/%s is a pre-created Subnet", bindingMap.Namespace, bindingMap.Spec.TargetSubnetName),
+				error:   fmt.Errorf("pre-created Subnet %s/%s cannot be a target Subnet", bindingMap.Namespace, bindingMap.Spec.TargetSubnetName),
+				retry:   false,
+			}
+		}
+	} else {
+		parentSubnetPaths, err = r.validateVpcSubnetsBySubnetSetCR(ctx, bindingMap.Namespace, bindingMap.Spec.TargetSubnetSetName)
 		if err != nil {
-			return nil, nil, err
+			return "", nil, err
 		}
-		return childSubnet, parentSubnets, nil
 	}
 
-	parentSubnets, err := r.validateVpcSubnetsBySubnetSetCR(ctx, bindingMap.Namespace, bindingMap.Spec.TargetSubnetSetName)
-	if err != nil {
-		return nil, nil, err
+	// If child Subnet is a pre-created Subnet, check if it is in the same vpc as parent Subnet
+	if childSubnetCR.Status.Shared {
+		childVpcPath, err := getVpcPath(childSubnetPath)
+		if err != nil {
+			return "", nil, err
+		}
+		parentVpcPath, err := getVpcPath(parentSubnetPaths[0])
+		if err != nil {
+			return "", nil, err
+		}
+		if childVpcPath != parentVpcPath {
+			return "", nil, &errorWithRetry{
+				message: fmt.Sprintf("Subnet %s and target Subnet %s are in different VPCs", childSubnetPath, parentSubnetPaths[0]),
+				retry:   false,
+				error:   fmt.Errorf("Subnet and target Subnet are in different VPCs"),
+			}
+		}
 	}
-	return childSubnet, parentSubnets, nil
+	return childSubnetPath, parentSubnetPaths, nil
 }
 
-func (r *Reconciler) validateVpcSubnetsBySubnetCR(ctx context.Context, namespace, name string, isTarget bool) ([]*model.VpcSubnet, *errorWithRetry) {
+func (r *Reconciler) validateVpcSubnetsBySubnetCR(ctx context.Context, namespace, name string, isTarget bool) ([]string, *v1alpha1.Subnet, *errorWithRetry) {
 	subnetCR := &v1alpha1.Subnet{}
 	subnetKey := types.NamespacedName{Namespace: namespace, Name: name}
 	// Check the Subnet CR existence.
 	err := r.Client.Get(ctx, subnetKey, subnetCR)
 	if err != nil {
 		log.Error(err, "Failed to get Subnet CR", "Subnet", subnetKey.String())
-		return nil, &errorWithRetry{
+		return nil, subnetCR, &errorWithRetry{
 			message: fmt.Sprintf("Unable to get Subnet CR %s", name),
 			retry:   false,
 			error:   fmt.Errorf("failed to get Subnet %s in Namespace %s with error: %v", name, namespace, err),
 		}
 	}
 
 	// Check the Subnet CR realization.
-	subnets := r.SubnetService.ListSubnetCreatedBySubnet(string(subnetCR.UID))
-	if len(subnets) == 0 {
-		log.Info("NSX VpcSubnets by subnet CR do not exist", "Subnet", subnetKey.String())
-		return nil, &errorWithRetry{
+	var subnetPaths []string
+	if subnetCR.Status.Shared {
+		realized := false
+		for _, con := range subnetCR.Status.Conditions {
+			if con.Type == v1alpha1.Ready && con.Status == corev1.ConditionTrue {
+				realized = true
+				break
+			}
+		}
+		if !realized {
+			return nil, subnetCR, &errorWithRetry{
+				message: fmt.Sprintf("Subnet CR %s is not realized on NSX", name),
+				retry:   false,
```

Comment:
shall we retry if the pre-created subnet is not realized?

Reply:
Subnetbindingmap has Subnet handler to requeue the Subnet if its readiness condition changes, so we do not need to retry here.

Reply:
good to know.